### ISSUE TITLE
How to create a pure client?

### ISSUE THREAD

**Issue by bolinfest** [ebkalderon/tower-lsp#217]
**tl;dr** I'm looking for something like the `createConnection()` function in `vscode-languageclient`:

https://github.com/microsoft/vscode-languageserver-node/blob/901fd8829ee0af8f457bc2a2b1ab5910abd669f6/client/src/common/client.ts#L169

I feel like a lot of the pieces are here to make a "pure LSP client," but I am unclear as to whether it's possible today. By a "pure client," I mean to assume you have an existing LSP server that can be treated as a black box. The caller is responsible for producing a bidrectional channel (often this is achieved by launching the LSP server and connecting stdin/stdio), which should be used to instantiate some sort of `Client` object. The [`vscode-languageclient`](https://www.npmjs.com/package/vscode-languageclient) npm module is a good example of a proper API+implementation for this:

https://github.com/microsoft/vscode-languageserver-node/blob/master/client/src/common/client.ts

As it stands, `tower_lsp::Client::new` is private and the only way to create one seems to be via `LspService::new()`, which requires returning an implementation of `tower_lsp::LanguageServer`. Assuming someone else has implemented the LSP server, why should I be required to implement `LanguageServer` in order to connect to it?

**Comment by ebkalderon** [comment]
Thanks for the question, @bolinfest! I think I may need to clarify the scope and purpose of this library a bit. As it stands, `tower-lsp` is more akin to `vscode-languageserver` rather than `vscode-languageclient`. It is intended as a scaffold for the creation of LSP servers and not clients, which is currently out of scope for this library. It assumes that a language client compatible with the editor you're targeting is already available or developed separately using the recommended toolkit, e.g. `vscode-language client`. The channel provided by `tower-lsp` allows for bidirectional communication insofar that it allows the server you are implementing to make requests back to the outside client and await responses. At present, `tower-lsp` provides no facilities for writing your own language client.

Although the development of language clients isn't currently on the radar, we can always discuss the merits of extending the existing library or shipping a companion library for writing clients in this issue thread. Hope this clarifies things for you!

**Comment by ebkalderon** [comment]
Closing this issue for now, since I believe the project scope described above summarizes the situation pretty well (this is a helper library for implementing language servers, not clients). However, anyone who's interested in this should feel free to reopen the issue for further discussion!
